rep-0151.rst

```diff
@@ -0,0 +1,293 @@
+
+Shebangs and reliance on the Python command
+'''''''''''''''''''''''''''''''''''''''''''
+Python scripts on UNIX systems typically have shebang lines written as:
+
+.. code-block: bash
+
+    #!/usr/bin/env python
+
+PEP 394 recommends distributed Python scripts to use either ``python2`` or
+``python3`` [7]_.
+The ``python`` command cannot be trusted to a specific Python version.
+On older ROS distros, scripts can continue to use ``python`` since they're known
+to work on those platforms.
+These shebangs must be rewritten to the specific version of Python supported.
+Packages can use the CMake macro ``catkin_install_python()`` to install Python
+scripts with rewritten shebangs.
+It will create a relay script in the ``devel`` space with a rewritten shebang.
+
+The same issue appears in scripts that call the ``python`` command directly.
+If they are Python scripts, they should invoke ``sys.executable``.
+Otherwise, they should invoke the specific version of Python they
+require, which means templating the script to invoke the Python interpreter
+found when the package was built.
+
+Dependencies and package.xml
+''''''''''''''''''''''''''''
+
+On platforms where the target version of Python is 2, the package.xml of a ROS
+package must refer to Python 2 dependencies, and when the target Python
+version is 3 it must refer to Python 3 dependencies.
+Packages which release from different branches for each ROS distro can replace
+rosdep keys that resolve to Python 2 dependencies with ones that resolve to
+Python 3 equivalents.
+Packages using the same code base for multiple ROS distros should instead use
+conditional dependencies as described in REP 149 [5]_.
+
+.. code-block: xml
+
+    <depend condition="$ROS_PYTHON_VERSION == '2'">python-numpy</depend>
+    <depend condition="$ROS_PYTHON_VERSION == '3'">python3-numpy</depend>
+
+If ``ROS_PYTHON_VERSION`` is relied upon at build time, such as when using
+``catkin_install_python()`` to rewrite shebangs, then the package must declare a
+``<buildtool_depend>`` on ``ros_environment``.
+Any ROS package which uses ``ROS_PYTHON_VERSION`` in a script intended to be
+run at runtime should add an ``<exec_depend>`` tag for ``ros_environment``.
+
+Making Python fixes available to downstream packages
+''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+Transitioning to Python 3 is expected to be a significant effort.
+Typically, ROS packages are tested using the ROS build farm; however, that
+will not be available until packages for the targeted Ubuntu distribution
+become available.
+Instead, a placeholder ``Noetic`` ``distribution.yaml`` will be made available in
+advance of the buildfarm availability.
+Maintainers should add ``source`` entries for their Noetic branches to this file
+to enable downstream users to use ``rosinstall_generator`` with the
+``--upstream-development`` flag to get Python 3 fixes.
+Instructions to build from source using Python 3 will be made available to
+the ROS community.
```

Comment:
Most of this paragraph seems obsolete now.

Reply:
Rewrote this and section below in 9c3048db45149849c8b95ab012c197aa2db7cd8d